### DISCUSSION TITLE
Local Notifications in .NET MAUI Essentials

### DISCUSSION THREAD

**Discussion by jfversluis** [dotnet/maui#4216] [Ideas]
There was a proposal [here](https://github.com/xamarin/Essentials/issues/238) in Xamarin.Essentials for local notifications. After it was initially turned down because it might be too specific to UI, there were some people who made a case for it and [an implementation was started](https://github.com/xamarin/Essentials/pull/739) to move the discussion forward.

Proposed API would look something like:

```csharp
class LocalNotifications {
  Task CancelAll();
  Task Cancel(int id);
  Task Send(LocalNotification notification);
  Task<IEnumerable<LocalNotification>> GetNotificationsAsync();
}

class LocalNotification {
  int Id { get; set; }
  string Title { get; set; }
  string Message { get; set; }
  IDictionary<string, string> Metadata { get; set; }
}
```

A plugin that already provides this and might act as inspiration: https://github.com/thudugala/Plugin.LocalNotification

What do people think? Should we add this to Essentials?

**Comment by aritchie** [comment]
This spec was fine for simple usage when I proposed it, but there are a lot of things feature wise that will be missing that many will find they need almost immediately after starting to use local notifications.
* Scheduled Notifications - this is almost a show stopper for local notifications.  It also doesn't come out of the box with Android
* Notification Actions/Replies - On Android, you really have to do this in local notifications
* Android has a ton of notifications customizations that are not "cross platform".  iOS has started to have their own customizations as well.
* Clicks Events - if the user taps on the notification, this event needs to be caught early in the app lifecycle to navigate properly

Other things to consider
* You need some form of more data store or serialization mechanism to be able to store the notifications for android.  Essentials doesn't currently have anything out of the box for this. 
* Long Term - Location Based Notifications - out of box for iOS, not Android.  You may be able to get around this by telling customers to use geofencing which essentials also doesn't offer

**Reply by carlfranklin** [reply]
I'd really like to use a SignalR message to trigger a notification. Seems like a match made in heaven.

**Reply by BobbyCannon** [reply]
Exactly what I was searching for when I got here. I want to send a signalR message but have it notify on the Android mobile device.

**Reply by aritchie** [reply]
SignalR connections won't stay connected when your app goes to the background.

**Reply by BobbyCannon** [reply]
What about running "foreground service"? I'm working on a local notification plug in right now. Maybe I should focus on the foreground service first.

**Reply by aritchie** [reply]
If you only care about Android, a foreground service could pull this off.  You could likely piggyback on the datasync foreground type, but Android updates may start cracking down on foreground services in the near future

**Comment by sander1095** [comment]
would really appreciate scheduled and cancellable notifications in this API. Granted, you have to mess about with the AlarmManager on Android but I think some extra features on this API would make it very useful.

**Comment by edgiardina** [comment]
This would be really cool. I would also like to pair this with iOS Background Fetch which I don't seem to see a lot of docs for MAUI on. My app has no push notifications powering it, instead we fetch and compare data periodically then local notify the user.

**Comment by gkarabin** [comment]
A low-feature, maintainable API would solve a number of simple use cases. I would think that there is room for something like the original proposal, a least common denominator. Once someone wants more differentiation they can look to investing in platform specifics.

That’s easy enough for me to write, since I’m not working on a contribution, but I think this is a case where the perfect is the enemy of the good enough, and this is a glaring hole.

**Comment by 99067718** [comment]
This is a really good idea, I hope this will get added.

**Comment by janseris** [comment]
When will this be available?
Does MAUI support some notifications at the moment? (not from external online service)

**Reply by liebki** [reply]
You have noticed, this is an complex situation? You can help by implementing this or you just have to wait so that someone else implements it.

**Reply by janseris** [reply]
Microsoft employees are supposed to implement it. I am not paid for that and noone else is than Microsoft employees. It's their responsibility.
Me and probably you as well are paid for developing business systems in our companies, not for developing MAUI framework. 
Who codes the business system instead of me if I'm wasting countless hours developing MAUI notifications which I know nothing about how to implement?
I have noticed this is a complex situation because the MAUI team (rebranded Xamarin) are thinking that someone else will implement it for them for free and have been waiting for this since 2018. It's embarrassing.
Firebase (their online notification service) can show notifications in MAUI Android. Why cannot MAUI Android show these notifications programatically as well?

**Reply by liebki** [reply]
You mean paying and no they don't have to implement anything. They implement it when they have time/capacity. They are human too, so be patient or implement it yourself if you have such a need for this feature (would benefit everybody).

Have a great day tho 🤓

**Reply by janseris** [reply]
I am sorry but that is economical nonsense. Our company is paying for Visual Studio, Windows Server, SQL Server and other Microsoft products. In exchange we receive a broken platform (Xamarin/MAUI) with 5 Microsoft devs who can never implement it fully in their lifetime and an opinion that we should develop MAUI ourselves (in "community")?
The 5 developers are of course humans and cannot do more work. The solution is that Microsoft needs more of these humans. Similar to our company when we have new/larger projects, we hire new people. I believe Microsoft can afford to pay new developers if our insignificiant company in eastern Europe can afford to.

**Reply by knocte** [reply]
> Our company is paying for Visual Studio, Windows Server, SQL Server and other Microsoft products.

Shouldn't this mean you're expected to receive support for VS, Windows and SQLServer? Or is there any clause in their license about opensource projects developed by Microsoft?

Look, I'd love if MS implemented this, but it's an opensource project and they don't get paid to develop it. But especially being opensource means that if you care enough, you can implement it yourself.

**Reply by janseris** [reply]
_Open source_ just means that everyone can see the source and maybe can contribute under very specific circumstances. This doesn't mean that it's never supposed to be finished or that MAUI/Xamarin team is not paid for developing MAUI.
> they don't get paid to develop it. 

I believe you are wrong. 
A lot of MAUI team people seen on GitHub are regular Microsoft employees.
Just check Microsoft MAUI YouTube videos.
Example:

@davidortinau David Ortinau is a "Principal Product Manager for .NET at Microsoft, focused on .NET MAUI. A .NET developer since 2002"
@maddymontaquila Maddy Montaquila - "Program Manager, Mobile Dev Tools at Microsoft "
@jamesmontemagno  James Montemagno -  "I work at Microsoft "
@Redth Jonathan Dick - "NET MAUI Team / Engineering Lead Principal Software Engineering Manager .NET MAUI @ Microsoft"
@jfversluis Gerald Versluis - "Software Engineer at microsoft"
@mattleibow  Matthew Leibowitz probably as well (I can't tell it directly)
@Eilon Eilon Lipton - has Microsoft work symbol only on his GitHub account

**Reply by knocte** [reply]
> I believe you are wrong.
> A lot of MAUI team people seen on GitHub are regular Microsoft employees.

I didn't say that Microsoft developers don't get paid to develop MAUI. I said that Microsoft doesn't get paid to develop MAUI.

**Reply by knocte** [reply]
> Open source just means that everyone can see the source

Woah boy you're so wrong.

**Reply by janseris** [reply]
> > I believe you are wrong.
> > A lot of MAUI team people seen on GitHub are regular Microsoft employees.
> 
> I didn't say that Microsoft developers don't get paid to develop MAUI. I said that Microsoft doesn't get paid to develop MAUI.

Well, Microsoft "doesn't get paid" directly to develop WinForms, Blazor, WPF, EF Core, .NET and all other platforms and frameworks. It's indirect business indeed. 
The platform is not sold to customers directly. 
But if you watch .NET development team, EF Core development team or ASP.NET development team, they are working hard, on time, bringing a lot of features which are not bugged and which work according to documentation because they have sufficient amount of people. 
On the other hand MAUI is the very opposite. MAUI team only labels new issues with colors because there are too few people. 
I have already been contacting the team sice April 2022 when it became evident that the issues are never to be resolved in current tempo. Maddy Montaquila explicitly asked for feedback in one of the MAUI YouTube presentations. I provided a feedback that there is a very large lack of devs and the issues are unmaintainable and I received a response that everything is fine with developer capacities. That was hillarious.

**Reply by liebki** [reply]
MAUI is literally a whole other thing, they try to bring multi-platform to TVs, Desktop, Phones etc using ONE framework. This is a huge amount of work, I don't think you can compare this to things like forms.

**Reply by janseris** [reply]
You are right. What I am just trying to say that MAUI is heavily underinvested in terms of developers and it's not changing and it looks unmantainable when you see the issues (+more and more to come). And one other thing is the quality of the already released MAUI technology and inconsistence with documentation. But that is probably the same cause - too few developers.
When we try to use any part of MAUI, we immediately are finding bugs, inconsistences with documentation (mostly that parts claimed to be working aren't working) and reporting loads of bugs with little perspective to be fixed (most of them haven't been fully fixed (most have been partially) since 6 months ago when we started with MAUI).
Using such technology as a client is very frustrating and that is why I am actively trying my best to let the team know and change because that's all I can do and hope for.

When you compare to other C# technologies by Microsoft, this is a rare failure because other products are supported and developed very well (well to post complete info, I read some unfortunate info about WPF being now supported by an outsourced team and not doing a quality work because they are lost in the technology).

**Reply by liebki** [reply]
For my things, I can say its doing its job. I can't talk about professional company stuff tho. I think with enough manpower and will, MAUI will be a fantastic thing.

**Reply by aritchie** [reply]
Wow... this thread exploded.  The level of entitlement that you can demand Microsoft do this is in full epic proportion as well.

A few reasons why Microsoft does not HAVE to implement this:
1) MAUI = Microsoft Application USER INTERFACE.  Notifications are not a render UI element.  They are a function of the platform
2) React Native & Flutter don't offer notifications out of the box.  It's from their corresponding ecosystems.
3) For simple notifications, you can implement this yourself in a few lines of code.
4) Essentials is basically bonus icing that isn't offered in many other xplat solutions or at least in the capacity that Essentials offers.
5) Learn to be thankful for what you do get for free in MAUI.  There's always flutter, react native, ionic if you think that will supply you better.  Pick your poison.

The under resourcing of the MAUI team in a github enhancement request, is ultimately noise and takes away from the justification for the feature.  It also takes the actual team MORE TIME to shovel through this crap.  This isn't the place for these types of comments.

For those looking for this particular feature.  The ecosystem has some great options.  I have an alpha version of Shiny.NET (https://github.com/shinyorg/shiny) that is compatible with .NET MAUI that offers many functions not available in MAUI Essentials including local & push notifications.

**Reply by liebki** [reply]
> Wow... this thread exploded.  The level of entitlement that you can demand Microsoft do this is in full epic proportion as well.
> 
> 
> 
> A few reasons why Microsoft does not HAVE to implement this:
> 
> 1) MAUI = Microsoft Application USER INTERFACE.  Notifications are not a render UI element.  They are a function of the platform
> 
> 2) React Native & Flutter don't offer notifications out of the box.  It's from their corresponding ecosystems.
> 
> 3) For simple notifications, you can implement this yourself in a few lines of code.
> 
> 4) Essentials is basically bonus icing that isn't offered in many other xplat solutions or at least in the capacity that Essentials offers.
> 
> 5) Learn to be thankful for what you do get for free in MAUI.  There's always flutter, react native, ionic if you think that will supply you better.  Pick your poison.
> 
> 
> 
> The under resourcing of the MAUI team in a github enhancement request, is ultimately noise and takes away from the justification for the feature.  It also takes the actual team MORE TIME to shovel through this crap.  This isn't the place for these types of comments.
> 
> 
> 
> For those looking for this particular feature.  The ecosystem has some great options.  I have an alpha version of Shiny.NET (https://github.com/shinyorg/shiny) that is compatible with .NET MAUI that offers many functions not available in MAUI Essentials including local & push notifications.  
> 
> 
> 
> 

Thank you for this

**Reply by janseris** [reply]
> Wow... this thread exploded. The level of entitlement that you can demand Microsoft do this is in full epic proportion as well.
> 
> A few reasons why Microsoft does not HAVE to implement this:
> 
>     1. MAUI = Microsoft Application USER INTERFACE.  Notifications are not a render UI element.  They are a function of the platform
> 
>     2. React Native & Flutter don't offer notifications out of the box.  It's from their corresponding ecosystems.
> 
>     3. For simple notifications, you can implement this yourself in a few lines of code.
> 
>     4. Essentials is basically bonus icing that isn't offered in many other xplat solutions or at least in the capacity that Essentials offers.
> 
>     5. Learn to be thankful for what you do get for free in MAUI.  There's always flutter, react native, ionic if you think that will supply you better.  Pick your poison.
> 
> 
> The under resourcing of the MAUI team in a github enhancement request, is ultimately noise and takes away from the justification for the feature. It also takes the actual team MORE TIME to shovel through this crap. This isn't the place for these types of comments.
> 
> For those looking for this particular feature. The ecosystem has some great options. I have an alpha version of Shiny.NET (https://github.com/shinyorg/shiny) that is compatible with .NET MAUI that offers many functions not available in MAUI Essentials including local & push notifications.

Thank you for link, nice to hear about that. I will take a look.
P.S.: MAUI being UI-only is only partial truth. 
MAUI contains also Maui.Graphics (since some of the last RCs) and Xamarin.Essentials (since Preview 11?).
 These are more non-UI than UI. Can't say how much code they make but I think it might be even 50 % of MAUI. 
 They allow access to device capabilities. This is an important part of MAUI which doesn't work correctly. 
 I can replace UI by paid UI libraries from third party vendors but these internals cannot be replaced. Notifications would be part of these. 
Remember when MAUI was released and almost no image manipulation from Maui.Graphics worked (half of the functions were throwing NotImplementedException and the other half threw about using a Disposed object) and the Windows implementation wasn't even a public class? That's what I am talking about.

>3. For simple notifications, you can implement this yourself in a few lines of code.

If so, could you guide me a bit of a hint? I'd like to try but I have no idea where to start. I don't know any Android development. 
MAUI is supposed to abstract programmer from the platform. That's why we started considering it and hoped that it would do having only experience in Windows desktop development.

**Comment by ComptonAlvaro** [comment]
I would like to see local notifications in the essentials, similar like the plugin you comment as inspiration.

One thing I would like to can do it is to handle acctions, for example, if I run a foreground service periodically as reminders, the user could tap a button in the notification to stop the reminder because he considerates as notified.
